### PR TITLE
entproto/cmd/protoc-gen-entgrpc: replace toProto func gen with template

### PR DESCRIPTION
**Pull request by rotemtam**
Following recently merged changes from #107, this PR replaces to `toProto` function generation with a `text/template` based approach.

The generated output has way too many new lines, I will fix this before merging.

### CODE REVIEW

**Comment by rotemtam** [comment]
cc @tarrencev

File: entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    // toProto{{ .EntType.Name }} transforms the ent type to the pb type
+    func toProto{{ .EntType.Name }}(e *{{ .EntPackage.Ident .EntType.Name | ident }}) (*{{ .EntType.Name }}, error) {
+        v := &{{ .EntType.Name }}{}
+        {{ range .FieldMap.Fields }}
+            {{- $varName := camel .EntField.StructField -}}
+            {{- $id := print "e." .EntField.StructField -}}
+            {{ template "field_to_proto" dict "Field" . "VarName" $varName "Ident" $id }}
+            v.{{ .PbStructField }} = {{ $varName }}
+
+        {{ end }}
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl:12]:
```suggestion
        {{- range .FieldMap.Fields }}
            {{- $varName := camel .EntField.StructField -}}
            {{- $id := print "e." .EntField.StructField -}}
            {{ template "field_to_proto" dict "Field" . "VarName" $varName "Ident" $id }}
            v.{{ .PbStructField }} = {{ $varName }}
        {{- end }}
```

Please check this before landing.

File: entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                }
+            {{ else }}
+                for _, edg := range e.Edges.{{ $name }} {
+                    {{ template "field_to_proto" dict "Field" . "VarName" $varName "Ident" $id }}
+                    v.{{ $name }} = append(v.{{ $name }}, &{{ .EntEdge.Type.Name }}{
+                        {{ .EdgeIDPbStructField }}: {{ $varName }},
+                    })
+                }
+            {{ end }}
+        {{ end }}
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl:32]:
```suggestion
        {{- range .FieldMap.Edges }}
            {{- $varName := camel .EntEdge.Type.ID.StructField -}}
            {{- $id := print "edg." .EntEdge.Type.ID.StructField -}}
            {{- $name := .EntEdge.StructField -}}
            {{- if .EntEdge.Unique }}
                if edg := e.Edges.{{ $name }}; edg != nil {
                    {{ template "field_to_proto" dict "Field" . "VarName" $varName "Ident" $id }}
                    v.{{ $name }} = &{{ .EntEdge.Type.Name }}{
                        {{ .EdgeIDPbStructField }}: {{ $varName }},
                    }
                }
            {{- else }}
                for _, edg := range e.Edges.{{ $name }} {
                    {{ template "field_to_proto" dict "Field" . "VarName" $varName "Ident" $id }}
                    v.{{ $name }} = append(v.{{ $name }}, &{{ .EntEdge.Type.Name }}{
                        {{ .EdgeIDPbStructField }}: {{ $varName }},
                    })
                }
            {{- end }}
        {{- end }}
```

File: entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        {{ .VarName }}Value, err := {{$id}}.Value()
+        if err != nil {
+            return nil, err
+        }
+        {{ .VarName }}Typed, ok := {{ .VarName }}Value.({{ $conv.ToProtoValuer }})
+        if !ok {
+            return nil, {{ qualify "errors" "New" }}("casting value to {{ $conv.ToProtoValuer }}")
+        }
+        {{ .VarName}} := {{ ident $conv.ToProtoConstructor }}({{ .VarName}}Typed)
+    {{- else if $conv.ToProtoValuer -}}
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl:58]:
```suggestion
    {{- if $conv.ToProtoConversion }}
        {{- $id = print $conv.ToProtoConversion "(" $id ")" }}
    {{- end -}}
    {{- if $conv.ToEntMarshallerConstructor.GoName }}
        {{ .VarName }}, err := {{ $id }}.MarshalBinary()
        if err != nil {
            return nil, err
        }
    {{- else if and $conv.ToProtoValuer $conv.ToProtoConstructor.GoName }}
        {{ .VarName }}Value, err := {{$id}}.Value()
        if err != nil {
            return nil, err
        }
        {{ .VarName }}Typed, ok := {{ .VarName }}Value.({{ $conv.ToProtoValuer }})
        if !ok {
            return nil, {{ qualify "errors" "New" }}("casting value to {{ $conv.ToProtoValuer }}")
        }
        {{ .VarName}} := {{ ident $conv.ToProtoConstructor }}({{ .VarName}}Typed)
    {{- else if $conv.ToProtoValuer }}
```

File: entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        {{ .VarName }}, ok := {{ .VarName }}Value.({{ $conv.ToProtoValuer }})
+        if !ok {
+            return nil, {{ qualify "errors" "New" }}("casting value to {{ $conv.ToProtoValuer }}")
+        }
+    {{- else if $conv.ToProtoConstructor.GoName -}}
+        {{ .VarName }} := {{ ident $conv.ToProtoConstructor }}({{ $id }})
+    {{- else -}}
+        {{ .VarName }} := {{ $id }}
+    {{- end -}}
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/to_proto.tmpl:71]:
```suggestion
    {{- else if $conv.ToProtoConstructor.GoName }}
        {{ .VarName }} := {{ ident $conv.ToProtoConstructor }}({{ $id }})
    {{- else }}
        {{ .VarName }} := {{ $id }}
    {{- end }}
```

**Review by a8m — APPROVED** [review]

**Comment by tarrencev** [comment]
One issue i encountered, that I'll mention here but we can solve later, is ent entity names can conflict with go reserved names. For example, an entity can be called `for`. We should add a prefix/suffix to generated var names to avoid these collisions.

**Comment by a8m** [comment]
> One issue i encountered, that I'll mention here but we can solve later, is ent entity names can conflict with go reserved names. For example, an entity can be called `for`. We should add a prefix/suffix to generated var names to avoid these collisions.

Can you share an example, because afaik, it’s not possible to name schema types this way:
https://github.com/ent/ent/blob/master/entc/gen/type.go#L796

**Comment by tarrencev** [comment]
> > One issue i encountered, that I'll mention here but we can solve later, is ent entity names can conflict with go reserved names. For example, an entity can be called `for`. We should add a prefix/suffix to generated var names to avoid these collisions.
> 
> Can you share an example? Because afaik, it’s not possible to name schema types this way:
> https://github.com/ent/ent/blob/master/entc/gen/type.go#L796

I have a schema that compiles with
```
field.Int("for").
			GoType(BigInt{}).
			SchemaType(BigIntSchemaType).
			DefaultFunc(func() BigInt {
				return NewBigInt(0)
			}),
```

**Comment by a8m** [comment]
> > > One issue i encountered, that I'll mention here but we can solve later, is ent entity names can conflict with go reserved names. For example, an entity can be called `for`. We should add a prefix/suffix to generated var names to avoid these collisions.
> > 
> > 
> > Can you share an example? Because afaik, it’s not possible to name schema types this way:
> > https://github.com/ent/ent/blob/master/entc/gen/type.go#L796
> 
> Hmm interesting. I have a schema that compiles with
> 
> ```
> field.Int("for").
>   GoType(BigInt{}).
>   SchemaType(BigIntSchemaType).
>   DefaultFunc(func() BigInt {
>     return NewBigInt(0)
>   }),
> ```

You're right. That's expected. I thought you meant the entity-type, but fields can be named as Go keywords.

**Comment by rotemtam** [comment]
> One issue i encountered, that I'll mention here but we can solve later, is ent entity names can conflict with go reserved names. For example, an entity can be called `for`. We should add a prefix/suffix to generated var names to avoid these collisions.

@tarrencev , did you encounter this issue in relation to this PR or is this a general comment?

**Comment by tarrencev** [comment]
> > One issue i encountered, that I'll mention here but we can solve later, is ent entity names can conflict with go reserved names. For example, an entity can be called `for`. We should add a prefix/suffix to generated var names to avoid these collisions.
> 
> 
> 
> @tarrencev , did you encounter this issue in relation to this PR or is this a general comment? 

Sorry i was unclear. This is an issue that was introduced by the value scanner work

**Pull request merged**